src/Typedefs/Test/JSONFormatTests.idr

```diff
@@ -0,0 +1,169 @@
+module Typedefs.Test.JSONFormatTests
+
+import Typedefs.Typedefs
+import Typedefs.Names
+import Typedefs.TermParse
+import Typedefs.TermWrite
+
+import Data.Vect
+import Data.Bytes
+import Data.ByteArray
+import TParsec
+import Specdris.Spec
+
+import Language.JSON
+
+%access public export
+%default total
+
+roundtrip1 : (td : TDefR 0) -> Ty [] td -> JSONM (Ty [] td)
+roundtrip1 td x = deserialiseJSON td [] $ serialiseJSON [] [] td x
+
+shouldBeRoundtrip1 : (td : TDefR 0) -> (Show (Ty [] td), Eq (Ty [] td)) => Ty [] td -> SpecResult
+shouldBeRoundtrip1 td term = (roundtrip1 td term) `shouldBe` (pure term)
+
+roundtrip2 : (td : TDefR 0) -> JSON -> JSONM JSON
+roundtrip2 td x = serialiseJSON [] [] td <$> deserialiseJSON td [] x
+
+Eq JSON where
+  (JNumber a) == (JNumber b) = a == b
+  JNull == JNull = True
+  (JBoolean a) == (JBoolean b) =  a == b
+  (JString a) == (JString b) = a == b
+  (JArray a) == (JArray b) = assert_total $ a == b
+  (JObject a) == (JObject b) = assert_total $ a == b
+  _ == _ = False
+
+-- helper functions to keep me sane while writing the tests
+jpair : (a, b : JSON) -> JSON
+jpair a b = JObject [("_0", a), ("_1", b)]
+
+jright : JSON -> JSON
+jright a = JObject [("_1", a)]
+
+jleft : JSON -> JSON
+jleft a = JObject [("_0", a)]
+
+jinn : JSON -> JSON
+jinn a = JObject [("inn", a)]
+
+junit : JSON
+junit = JObject []
+
+testSuite : IO ()
+testSuite = spec $ do
+
+  describe "TermWrite" $ do
+
+    it "serialise unit" $
+      (serialiseJSON [] [] T1 ()) `shouldBe` (JObject [])
+
+    it "serialise sum" $
+      (serialiseJSON [] [] (TSum [T1, T1]) (Left ())) `shouldBe` (JObject [("_0", JObject [])])
```

Comment:
`jleft junit`? Seems a shame to introduce this little DSL but not use it consistently

Reply:
Yes, I wrote this before I added junit and jleft/jright. Ill change this